src/Microsoft.ML.Data/DataLoadSave/Binary/Codecs.cs

```diff
@@ -1276,7 +1272,116 @@ private bool GetKeyCodec(Stream definitionStream, out IValueCodec codec)
+
+            private readonly CodecFactory _factory;
+            private readonly KeyType _type;
+            // We rely on a more basic value codec to do the actual saving and loading.
+            private readonly IValueCodec<T> _innerCodec;
+
+            public string LoadName { get { return "Key2"; } }
+
+            public ColumnType Type { get { return _type; } }
+
+            public KeyCodec(CodecFactory factory, KeyType type, IValueCodec<T> innerCodec)
+            {
+                Contracts.AssertValue(factory);
+                Contracts.AssertValue(type);
+                Contracts.AssertValue(innerCodec);
+                Contracts.Assert(type.RawType == typeof(T));
+                Contracts.Assert(innerCodec.Type.RawKind == type.RawKind);
+                _factory = factory;
+                _type = type;
+                _innerCodec = innerCodec;
+            }
+
+            public int WriteParameterization(Stream stream)
+            {
+                int total = _factory.WriteCodec(stream, _innerCodec);
+                using (BinaryWriter writer = _factory.OpenBinaryWriter(stream))
+                {
+                    writer.Write(_type.Count);
+                    total += sizeof(ulong);
+                }
+                return total;
+            }
+
+            // REVIEW: There is something a little bit troubling here. If someone, say,
+            // produces a column on KeyType(I4, 4) and then returns 10 as a value in
+            // that column, that's obviously a violation of the type, and lots of things
+            // downstream may complain, but it is a "valid" cursor in that it produces values
+            // and does not throw. So from that perspective of the codecs and their users being
+            // common and indifferent carriers, it's not clear tha these codecs should take on the
+            // responsibility for validating the input. On the *other* hand, if we know that we
+            // wrote valid data, when reading it back from a stream should we not take advantage
+            // of this, to validate the correctness of the decoding? On the other other hand, is
+            // validating the correctness for the decoding of things like streams any less urgent?
+
+            public IValueWriter<T> OpenWriter(Stream stream)
+            {
+                return _innerCodec.OpenWriter(stream);
+            }
+
+            public IValueReader<T> OpenReader(Stream stream, int items)
+            {
+                return _innerCodec.OpenReader(stream, items);
+            }
+        }
+
+        private bool GetKeyCodec(Stream definitionStream, out IValueCodec codec)
+        {
+            // The first value in the definition stream will be the internal codec.
+            IValueCodec innerCodec;
+            if (!TryReadCodec(definitionStream, out innerCodec))
+            {
+                codec = default(IValueCodec);
```

Comment:
Maybe this can work.
```suggestion
                codec = default;
```